### ISSUE TITLE
[C API] Convert a few stdlib extensions to the limited C API (PEP 384)

### ISSUE THREAD

**Issue by vstinner** [python/cpython#85283]
**Discussion:** https://discuss.python.org/t/use-the-limited-c-api-for-some-of-our-stdlib-c-extensions/32465

BPO | [41111](https://bugs.python.org/issue41111)
--- | :---
Nosy | @freddrake, @smontanaro, @rhettinger, @vstinner, @tiran, @encukou, @pmp-p, @ericsnowcurrently, @corona10, @shihai1991
PRs | <li>python/cpython#25115</li><li>python/cpython#25151</li><li>python/cpython#25180</li><li>python/cpython#26080</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2020-06-25.08:54:20.867>
labels = ['extension-modules', 'expert-C-API', '3.11']
title = '[C API] Convert a few stdlib extensions to the limited C API (PEP 384)'
updated_at = <Date 2022-02-17.16:21:09.570>
user = 'https://github.com/vstinner'
```

bugs.python.org fields:
```python
activity = <Date 2022-02-17.16:21:09.570>
actor = 'eric.snow'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Extension Modules', 'C API']
creation = <Date 2020-06-25.08:54:20.867>
creator = 'vstinner'
dependencies = []
files = []
hgrepos = []
issue_num = 41111
keywords = ['patch']
message_count = 27.0
messages = ['372307', '372309', '372310', '372311', '372315', '372316', '372321', '372324', '372325', '372327', '372332', '372386', '376506', '376507', '380415', '389925', '389959', '390068', '390159', '390161', '390179', '390326', '390333', '390389', '393541', '393635', '393640']
nosy_count = 11.0
nosy_names = ['fdrake', 'skip.montanaro', 'rhettinger', 'gustavo', 'vstinner', 'christian.heimes', 'petr.viktorin', 'pmpp', 'eric.snow', 'corona10', 'shihai1991']
pr_nums = ['25115', '25151', '25180', '26080']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue41111'
versions = ['Python 3.11']
```

</p></details>

### Linked PRs
* gh-108500
* gh-108501
* gh-108506
* gh-108570
* gh-108571
* gh-108572
* gh-108573
* gh-108577
* gh-108965
* gh-110711
* gh-110725
* gh-110954
* gh-110955
* gh-110967
* gh-110976
* gh-110978
* gh-110989
* gh-111005
* gh-111008
* gh-111010
* gh-111087
* gh-116611
* gh-116791
* gh-116841
* gh-116927
* gh-116928
* gh-117125
* gh-117321

**Comment by vstinner** [comment]
I just added a partial support for the limited C API to Argument Clinic (issue #108494, PR #108495). We can now reconsider converting some stdlib C extensions to the limited C API. So I reopen the issue.

**Comment by vstinner** [comment]
I wrote 3 PRs to convert 3 extensions to the limited C API:

* _uuid: PR #108506, functions have no argument, no impact on performance
* _testmultiphase: PR #108501. PyArg_ParseTuple() is used directly: Argument Clinic is not used.
* _statistics: PR #108500. This change makes the only function 2x slower because PyArg_ParseTuple() and METH_VARARGS are slower than AC optimized code with METH_FASTCALL.

I suggest to leave _statistics unchanged for now. It's not a good target to start the work on this topic.

**Comment by vstinner** [comment]
One issue is that the TraceRefs build is incompatible with the limited C API 😬

**Comment by erlend-aasland** [comment]
> To start the work on this issue, I propose by converting the `_stat` extension since the only change that it requires is to define the `Py_LIMITED_API` macro: PR #108573. I hesitate between documenting this change between the "C API" section and the "Build" section. Moreover, I'm not sure if it should be documented as a "Build Changes" in What's New in Python 3.13.

I think _Build_ is a better fit than _C API_, though I agree it is not optimal.

**Comment by vstinner** [comment]
> One issue is that the TraceRefs build is incompatible with the limited C API 😬

I created issue #108634 for that.

**Comment by vstinner** [comment]
Extensions which look interesting to be converted to the limited C API:

* ``_bisect``: use ``PyObject_CallOneArg()``, ``_PyObject_CallMethod()`` which can be replaced with ``PyObject_CallMethodOneArg()``. Argument Clinic uses ``_Py_convert_optional_to_ssize_t()``, it is going to be fixed.
* ``fcntl``: need ``PySys_Audit()``, Argument Clinic uses private ``_PyLong_FileDescriptor_Converter()``
* ``syslog``: need ``PySys_Audit()``, ``Py_XSETREF()`` can be replaced with ``Py_XDECREF()``, ``PyUnicode_AsUTF8()`` should be replaced with ``PyUnicode_AsUTF8String()`` (replace ``S_ident_o`` global with ``S_ident_utf8``).
* ``md5``: replace ``_PyType_GetModuleState()`` with ``PyType_GetModuleState()``. ``_Py_strhex()`` can be rewritten.
* ``pwd``: need ``PyMem_RawRealloc()``, ``posixmodule.h`` should not exclude the limited C API.
* ``grp``: need ``PyMem_RawRealloc()``, replace ``PyStructSequence_SET_ITEM()`` with ``PyStructSequence_SetItem()``, ``posixmodule.h`` should not exclude the limited C API.
* ``resource``: need ``PySys_Audit()``, replace ``PyStructSequence_SET_ITEM()`` with ``PyStructSequence_SetItem()``.

md5: replace ``_Py_strhex()``:

```c
    const char *hexdigits = "0123456789abcdef";
    char digest_hex[MD5_DIGESTSIZE * 2];
    char *str = digest_hex;
    for (size_t i=0; i < MD5_DIGESTSIZE; i++) {
        unsigned char byte = digest[i];
        *str++ = hexdigits[byte >> 4];
        *str++ = hexdigits[byte & 0x0f];
    }
    return PyUnicode_FromStringAndSize(digest_hex, sizeof(digest_hex));
```

**Comment by vstinner** [comment]
C extensions which would be interesting to build with the limited C API:

* Modules/_blake2/blake2module.c -- `blake2_exec()` needs `PyType_GetDict()` which is excluded from the limited C API
* Modules/_ctypes/_ctypes_test.c
* Modules/_xxtestfuzz/_xxtestfuzz.c
* Modules/_dbmmodule.c
* Modules/_gdbmmodule.c
* Modules/_statisticsmodule.c
* Modules/_testbuffer.c
* Modules/_uuidmodule.c
* Modules/syslogmodule.c
* Modules/xxmodule.c
* Modules/fcntlmodule.c
* Modules/errnomodule.c
* Modules/overlapped.c
* Modules/resource.c
* Modules/xxsubtype.c
* Modules/_scproxy.c
* Modules/_testclinic_limited.c
* Modules/_testimportmultiple.c
* Modules/grpmodule.c
* Modules/pwdmodule.c
* Modules/_stat.c
* Modules/_testcapimodule.c
* Modules/termios.c
* PC/winsound.c

Others:

* Modules/_bisectmodule.c: `_PyObject_CallMethod()` should be replaced with `PyObject_CallMethodObjArgs()`
* Modules/_multiprocessing/posixshmem.c - use `PyUnicode_AsUTF8()` (non-limited C API)
* Modules/_multiprocessing/multiprocessing.c - use `_PyOS_IsMainThread()` of `pycore_signal.h`
* Modules/_contextvarsmodule.c -- `PyContext_CopyCurrent()` is non-limited C API
* Modules/_csv.c -- use PyUnicode_4BYTE_KIND, Py_XSETREF(), etc.
* Modules/_curses_panel.c -- Function _curses_panel.panel.bottom cannot use limited C API

**Comment by vstinner** [comment]
Annoying API: ``type->tp_name`` is not accessible in the limited C API, and replacement APIs are not trivial to use. Example from my winsound change:

```diff
-        PyErr_Format(PyExc_TypeError,
-                     "'sound' must be str, os.PathLike, or None, not '%s'",
-                     Py_TYPE(sound)->tp_name);
+        PyObject *type_name = PyType_GetQualName(Py_TYPE(sound));
+        if (type_name != NULL) {
+            PyErr_Format(PyExc_TypeError,
+                         "'sound' must be str, os.PathLike, or None, not %S",
+                         type_name);
+            Py_DECREF(type_name);
+        }
```

**Comment by vstinner** [comment]
Oh. Since commit 37bd8726b87c1c41f58cef031f5b6d5f900d265b (Build errno and _ctypes_test with limited C API), build fails on Windows in Azure Pipelines: https://dev.azure.com/Python/cpython/_build/results?buildId=140935&view=logs&j=c8a71634-e5ec-54a0-3958-760f4148b765&t=03176981-a846-5848-f289-6e888069ab02

```
  _ctypes_test.c
LINK : fatal error LNK1104: cannot open file 'python3.lib' [D:\a\1\s\PCbuild\_ctypes_test.vcxproj]
```

On GitHub Actions, the build is fine on Windows x86 and Windows x64 jobs.

Same on buildbot: no build issue on Windows buildbots.

**Comment by vstinner** [comment]
Oh. On my PR #110955, the Windows build just passed. It's even more strange. But on the PR, only "Pre-build" and "Ubuntu" jobs ran: https://dev.azure.com/Python/cpython/_build/results?buildId=140895&view=results

**Comment by vstinner** [comment]
I can also build Python locally on Windows: no build issue.

**Comment by vstinner** [comment]
Ah wait, I reproduced the issue: the build fails when Python is built in release mode, without `-d` option.

**Comment by vstinner** [comment]
> Oh. Since commit https://github.com/python/cpython/commit/37bd8726b87c1c41f58cef031f5b6d5f900d265b (Build errno and _ctypes_test with limited C API), build fails on Windows in Azure Pipelines

It should be fixed by https://github.com/python/cpython/pull/111005 Job currently running: https://dev.azure.com/Python/cpython/_build/results?buildId=140979&view=results

**Comment by vstinner** [comment]
> Modules/_blake2/blake2module.c -- blake2_exec() needs PyType_GetDict() which is excluded from the limited C API

I tried to replace PyDict_SetItemString() with PyObject_SetAttrString(), but the type is immutable. The exec function sets attributes of an immutable type, hum.

**Comment by vstinner** [comment]
Workaround for the lack of PyUnicode_AsUTF8() in the stable ABI: replace PyUnicode_AsUTF8(obj) with PyUnicode_AsUTF8AndSize(obj, NULL) and ignore the size. PyUnicode_AsUTF8() already ignores embedded null character, so it doesn't change anything.

**Comment by vstinner** [comment]
syslog is blocked by this check, PyInterpreterState_Main() doesn't existing in the limited C API:

```c
static inline int
is_main_interpreter(void)
{
    return (PyInterpreterState_Get() == PyInterpreterState_Main());
}

...
    if (!is_main_interpreter()) {
        PyErr_SetString(PyExc_RuntimeError, "subinterpreter can't use syslog.openlog()");
        return NULL;
    }
```

**Comment by vstinner** [comment]
The following clinic.py code is not compatible with the limited C API:

```py
                return (f'PyErr_Format(PyExc_TypeError, '
                        f'"{{{{name}}}}() {displayname} must be %.50s, not %.50s", '
                        f'"{expected}", '
                        f'{{argname}} == Py_None ? "None" : Py_TYPE({{argname}})->tp_name);')
```

Using PyType_GetQualName()+PyUnicode_AsUTF8AndSize() is not easy because of the "None" code path, to display "None" instead of "NoneType".

**Comment by vstinner** [comment]
> syslog is blocked by this check, PyInterpreterState_Main() doesn't existing in the limited C API: (...)

I created issue gh-111088 "C API: Design an API to disallow creating more than one extension instance per Python process".

**Comment by vstinner** [comment]
The work is done. If there are more C extensions to convert, I may create new issues instead.

> The following clinic.py code is not compatible with the limited C API: (... tp_name ...)

I created https://github.com/python/cpython/issues/111696 for that.

**Comment by AraHaan** [comment]
It would be nice if ``PyObject_VectorcallDict`` could be in the limited C api as well as it is stable.

**Comment by vstinner** [comment]
> It would be nice if PyObject_VectorcallDict could be in the limited C api as well as it is stable.

This issue is closed. Please open a new issue with the rationale for adding `PyObject_VectorcallDict()` to the limited C API.

**Comment by vstinner** [comment]
@AraHaan: Why do you need `PyObject_VectorcallDict()`? Can you elaborate?

**Comment by AraHaan** [comment]
> @AraHaan: Why do you need `PyObject_VectorcallDict()`? Can you elaborate?

In the limited ABI their lacks a version of vector call that accepts ``args`` as a dict, and ``kwargs`` as a tuple instead of the confusing to use ``kwnames`` versions that I see no example usages of at all on the internet it seems.

**Comment by vstinner** [comment]
I created https://github.com/python/cpython/issues/120881: [C API] Add PyObject_VectorcallDict() to the limited C API.
